### PR TITLE
HTCONDOR-1589-internal-cgroup-v1

### PR DESCRIPTION
**Pull request by GregThain**
Move cgroup v1 support out of starter/procd and into daemon core.  Also remove libcgroup dependency.  This is because libcgroup isn't supported or packaged on all our of cgroup v1 platforms, or our cgroup v2 platforms that might need cgroup v1 if they are running inside a container.  Also, it sets us up to put shadows or other daemons inside their own cgroups.

This shouldn't be user visible, so not sure a version history entry is needed?

<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: docs/version-history/feature-versions-10-x.rst [hunk 1]
```diff
@@ -36,6 +36,11 @@ New Features:
   :ref:`users-manual/dagman-workflows:dag save point files`.
   :jira:`1636`
 
+- Linux Cgroup support has been redone in a way that doesn't depend on
+  using the procd.  There should be now user visible changes in
```
Comment (Todd-L-Miller) [inline docs/version-history/feature-versions-10-x.rst:40]:
```suggestion
  using the procd.  There should be no user visible changes in
```

File: src/condor_utils/proc_family_direct_cgroup_v1.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+// the libcgroup library.  This library was abandoned,
+// and only availble on platforms that support cgroup v1
+// natively.  Sometimes, we run on platforms like el9
+// that don't support cgroup v1, and thus don't have a
+// libcgroup avaible, but we are running inside a container,
+// and the *host's* cgroup version is what matters.  In 
+// this case we need to manipulate cgroupv1.
+//
+// So, we do it ourselves now, via raw writes to /sys/fs/cgropu
+// in the same way we do with cgroup v1.  See the
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.h:40]:
```suggestion
// in the same way we did via the procd.  See the 
```

??

File: src/condor_utils/proc_family_direct_cgroup_v1.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	bool quit(void(*)(void*, int, int),void*) { return false; }; // nothing to do here, only needed when there is a procd
+
+	// Have we seen an oom kill event in this cgroup;
+	bool has_been_oom_killed(pid_t pid);
+
+	// We don't need these, cgroups just works
+	bool track_family_via_environment(pid_t, PidEnvID&) {return true;}
+	bool track_family_via_login(pid_t, const char*) {return true;}
+	bool track_family_via_allocated_supplementary_group(pid_t, gid_t&) { return true; }
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.h:94]: All of the names in the class are terrible (not your fault, but what's the difference between a family and a subfamily?), but these _really_ seem like they ought to all return `false` when we _don't_ need them.

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+void fullyRemoveCgroup(const stdfs::path &absCgroup) {
+
+	// Common case, it shouldn't still exist, we're done
+	if (!stdfs::exists(absCgroup)) {
+		return;
+	}
+
+	// Otherwise, we need to depth-firstly remove any subdirs
+	for (auto const& entry: std::filesystem::directory_iterator{absCgroup})
+    {
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:57]: This seems unlikely to be what you intended.
```suggestion
	for (auto const& entry: std::filesystem::directory_iterator{absCgroup}) {
```
?

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+			int r = write(fd, buf.c_str(), strlen(buf.c_str()));
+			if (r < 0) {
+				dprintf(D_ALWAYS, "Error writing procid %d to %s: %s\n", pid, procs_filename.c_str(), strerror(errno));
+				close(fd);
+				return false;
+			} else {
+				dprintf(D_ALWAYS, "Moved process %d to cgroup %s\n", pid, absolute_cgroup.c_str());
+			}
+			close(fd);
+		}
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:117]: So if we fail to open `cgroup.procs` and therefore never actually assign the process to the cgroup, it's OK?  What am I missing, here?

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+		} else {
+			dprintf(D_ALWAYS, "Error setting cgroup memory limit of %lu in cgroup %s: %s\n", cgroup_memory_limit, memory_limits_path.c_str(), strerror(errno));
+		}
+	} else {
+		dprintf(D_FULLDEBUG, "ProcFamilyDirectCgroupV1 not setting any cgroup memory limits\n");
+	}
+
+	// Set cpu limits, if any
+	if (cgroup_cpu_shares > 0) {
+		// write memory limits
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:142]:
```suggestion
		// write CPU limits
```

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+				dprintf(D_ALWAYS, "Error reading system_usec field out of cpu.stat\n");
+				fclose(f);
+				return false;
+			}
+		}
+	}
+	fclose(f);
+
+	time_t wall_time = time(nullptr) - start_time;
+	usage.percent_cpu = double(user_usec + sys_usec) / double((wall_time * 1'000'000));
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:241]: This is so pretty.  *sniff*

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+		fclose(f);
+	}
+
+	// usage is in kbytes.  cgroups reports in bytes
+	usage.total_image_size = usage.total_resident_set_size = (memory_current_value / 1024);
+
+	// Sanity check if system is missing memory.peak file
+	if (memory_current_value > memory_peak_value) {
+		memory_peak_value = memory_current_value;
+	}
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:286]: Glad to see this.

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+		usage.max_image_size = memory_peak_value / 1024;
+	}
+
+	return true;
+}
+
+// Note that in cgroup v1, cgroup.procs contains only those processes in
+// this direct cgroup, and does not contain processes in any descendent
+// cgroup (except the / croup, which does).
+	bool
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:299]:
```suggestion
bool
```

... as long as I'm suggesting thirty billion changes already...

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+		close(fd);
+	} else {
+		dprintf(D_ALWAYS, "ProcFamilyDirectCgroupV1::continue_family error %d (%s) opening cgroup.freeze\n", errno, strerror(errno));
+		success = false;
+	}
+
+	return success;
+}
+
+	bool
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:383]:
```suggestion
bool
```

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+ProcFamilyDirectCgroupV1::kill_family(pid_t pid)
+{
+	std::string cgroup_name = cgroup_map[pid];
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:386]:
```suggestion

```
?

**Review by Todd-L-Miller — APPROVED** [review]
Big fan of `std::filesystem::path`, less of a fan of the custom abbreviation.  *shrug*

Anyway, some of these functions are really pretty.

File: src/condor_utils/proc_family_direct_cgroup_v1.h (hunk 3 above)
Comment (GregThain) [reply src/condor_utils/proc_family_direct_cgroup_v1.h:94]: DaemonCore::CreateProcess unconditionally calls all of these methods, and fails if they return false, so they have to return true, even though it is a lie.

Comment (Todd-L-Miller) [reply src/condor_utils/proc_family_direct_cgroup_v1.h:94]: Oh, ew.  OK.

**Review by Todd-L-Miller — APPROVED** [review]

**Review by Todd-L-Miller — APPROVED** [review]

**Review by Todd-L-Miller — APPROVED** [review]

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+			dprintf(D_ALWAYS, "Error setting cgroup cpu weight of %d in cgroup %s: %s\n", cgroup_cpu_shares, cpu_shares_path.c_str(), strerror(errno));
+		}
+	}
+
+	// Chown the leaf cgroup dirs to the non-root owner.  The control files wlil
+	// still be owned by root, so the job can't change them, but they can
+	// create sub-dirs, so that a glidein can manage the resources given in this
+	// cgroup
+	pid_t uid = get_user_uid();
+	pid_t gid = get_user_gid();;
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:163]:
```suggestion
	pid_t gid = get_user_gid();
```

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp [hunk 13]
```diff
@@ -0,0 +1,553 @@
...
+
+	// in a personal condor, uid/gid aren't returned properly.
+	if ((uid > 0) && (gid > 0)) {
+		for (auto &controller: controllers) {
+			int r = chown((cgroup_root_dir / controller / cgroup_name).c_str(), uid, gid);
+			if (r < 0) {
+				dprintf(D_FULLDEBUG, "Error chowning cgroup directory: %s to (%d.%d)\n", strerror(errno), uid, gid);
```
Comment (Todd-L-Miller) [inline src/condor_utils/proc_family_direct_cgroup_v1.cpp:170]: I guess this error will be detected when we try doing something later?

**Review by Todd-L-Miller — APPROVED** [review]
More minor questions but otherwise looks good.

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp (hunk 13 above)
Comment (GregThain) [reply src/condor_utils/proc_family_direct_cgroup_v1.cpp:170]: Yes -- at the worst it falls back to today's behaviour. And in the usual case where there isn't a glidein running in us, there's no difference.

File: src/condor_utils/proc_family_direct_cgroup_v1.cpp (hunk 5 above)
Comment (GregThain) [reply src/condor_utils/proc_family_direct_cgroup_v1.cpp:119]: Ah, good catch.  Change pushed.

**Review by Todd-L-Miller — APPROVED** [review]

**Pull request merged**